File: 1232-sum-of-mutated-array-closest-to-target/sum-of-mutated-array-closest-to-target.js
var findBestValue = function(arr, target) {
    let i = 0
    let j = Math.max(...arr)
    let minDiff = Infinity
    let result = 1

    function addSum(mid) {
        let sum = 0
        for (let val of arr) sum += val > mid ? mid: val
        return sum
    }

    while (i <= j) {
        let mid = Math.floor((i + j) / 2)
        let sum = addSum(mid)
        if (sum == target) return mid
        if (sum < target) i = mid + 1
        else j = mid - 1
        diff = Math.abs(target - sum)
        // if (diff < minDiff || diff == minDiff) {
        //     result = mid;
        //     minDiff = diff;
        // }
        if (diff < minDiff || diff == minDiff && mid < result) {
            result = mid;
            minDiff = diff;
        }
    }
    return result
};

File: 1232-sum-of-mutated-array-closest-to-target/sum-of-mutated-array-closest-to-target.py
class Solution:
    def findBestValue(self, arr: List[int], target: int) -> int:
        i = 0
        j = max(arr)
        min_diff = float("inf")
        result = float("inf")

        def addSum(num):
            total = 0
            
            for val in arr:
                total += num if val > num else val

            return total

        while (i <= j):
            mid = (i + j) // 2

            summ = addSum(mid)

            if (summ == target):
                return mid
            if (summ < target):
                i = mid + 1
            else:
                j = mid - 1

            diff = abs(summ - target)

            if (diff < min_diff or (diff == min_diff and mid < result)):
                result = mid
                min_diff = diff

        
        return result
        